foreshadow/core/base.py

```diff
@@ -0,0 +1,185 @@
+
+        """
+        transformer_list = [(smart.__name__, smart(), [col]) for col in X]
+        transformer_weights = None
+        return transformer_list, transformer_weights
+
+    def parallelize_smart_steps(self, X):
+        """Make self.steps for internal pipeline methods parallelized.
+
+        Takes self._steps passed at construction time and wraps each step
+        with ParallelProcessor to parallelize it across a DataFrame's columns.
+        Made possible sing get_transformer_list and get_transformer_weights
+        which must be implemented by the subclass.
+
+        Args:
+            X: DataFrame
+
+        """
+        steps = []
+        for step in self._steps:
+            name = step[0]
+            transformer = step[1]
+            transformer_list = self.get_transformer_list(transformer, X)
+            transformer_weights = self.get_transformer_weights(transformer, X)
+            steps.append(
+                (
+                    name,
+                    ParallelProcessor(
+                        transformer_list,
+                        transformer_weights=transformer_weights,
+                    ),
+                )
+            )
+        self.steps = steps
+
+    @abstractmethod
+    def get_transformer_list(self, transformer, X):
+        """Return valid transformer_list for ParallelProcessor.
+
+        Args:
+            transformer: SmartTransformer to use
+            X: input DataFrame
+
+        Returns:
+            ['name', 'transformer', ['cols']] format where each column
+            should be included exactly once.
+
+        .. # noqa: I202
+
+        """
+        pass
+
+    @abstractmethod
+    def get_transformer_weights(self, transformer, X):
+        """Return valid transformer_weights for ParallelProcessor.
+
+        Args:
+            transformer: SmartTransformer to use
+            X: input DataFrame
+
+        Returns:
+            {'transformer': weight} weighting. See FeatureUnion object.
+
+        .. # noqa: I202
+
+        """
+        pass
+
+    # main underlying fit(_transform)
+    # method called for fit, fit_transform, etc. Returns fit_transformed X.
+    def _fit(self, X, *args, **kwargs):
+        """Parallelize workflow then fit_transform data.
+
+        Standard sklearn Pipeline fit where each step in the pipeline is
+        parallelized using self.parallelize_smart_steps.
+
+        Args:
+            X: input DataFrame
+            *args: args to _fit
+            **kwargs: kwargs to _fit
+
+        Returns:
+            transformed data handled by Pipeline._fit
+
+        """
+        self.parallelize_smart_steps(X)
+        return super()._fit(X, *args, **kwargs)  # internally will iterate
+        # over steps in _fit.
+
+
+class SinglePreparerStep(PreparerStep):
+    def __init__(self, smart_transformer):
+        smart_transformer = [smart_transformer]
+        super().__init__(smart_transformer)
+        self.transformer = deepcopy(self._steps[0])
```

Comment:
This means that transformer will be "out of sync" with steps is that the intent here?